File: fitness_engine/models/nutrition.py
"""
Nutrition plan data models.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from enum import Enum

from ..utils.serialize import convert_for_json


class CalorieStrategy(str, Enum):
    DEFICIT = "deficit"            # cut
    SURPLUS = "surplus"            # bulk
    MAINTENANCE = "maintenance"    # maintenance
    RECOMP = "recomp"              # slight deficit
    REVERSE_DIET = "reverse_diet"  # transition


class RMRFormula(str, Enum):
    MIFFLIN_ST_JEOR = "mifflin_st_jeor"
    HARRIS_BENEDICT_ORIG = "harris_benedict_original"
    HARRIS_BENEDICT_REVISED = "harris_benedict_revised"
    CUNNINGHAM = "cunningham"           # Cunningham 1991: RMR = 500 + 22 × FFM
    KATCH_MCARDLE = "katch_mcardle"     # Katch-McArdle 1975: RMR = 370 + 21.6 × LBM


@dataclass
class RMRResult:
    formula: RMRFormula
    base_rmr_kcal: float
    metabolic_adaptation_factor: float = 1.0    # multiplier (e.g., 0.95 if in deficit)
    weight_reduced_factor: float = 1.0           # multiplier (e.g., 0.97 if >10% below max)
    adjusted_rmr_kcal: float = 0.0
    notes: list[str] = field(default_factory=list)


@dataclass
class TDEEResult:
    rmr_kcal: float
    activity_factor: float
    tdee_kcal: float
    adaptive_tdee_kcal: float | None = None    # if intake/weight logs provided
    final_tdee_kcal: float = 0.0
    notes: list[str] = field(default_factory=list)


@dataclass
class CalorieTargets:
    strategy: CalorieStrategy
    base_tdee_kcal: float
    rate_pct: float                                # weekly (cut) or monthly (bulk) % BW
    rate_label: str                                # human-readable
    calorie_delta_kcal: float                      # negative for cut, positive for bulk
    target_calories_kcal: float
    calorie_floor_applied: bool = False
    floor_kcal: int | None = None
    notes: list[str] = field(default_factory=list)

    def __post_init__(self):
        """Tier 3.32 fix: validate output ranges."""
        if math.isnan(self.target_calories_kcal) or self.target_calories_kcal <= 0:
            raise ValueError(
                f"CalorieTargets.target_calories_kcal must be positive; "
                f"got {self.target_calories_kcal}"
            )
        if math.isnan(self.base_tdee_kcal) or self.base_tdee_kcal <= 0:
            raise ValueError(
                f"CalorieTargets.base_tdee_kcal must be positive; got {self.base_tdee_kcal}"
            )


@dataclass
class MacroSplit:
    protein_g: float
    fat_g: float
    carb_g: float
    protein_pct: float
    fat_pct: float
    carb_pct: float
    protein_kcal: float
    fat_kcal: float
    carb_kcal: float
    notes: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        """Phase-6: validate macro consistency.

        - Percentages must sum to ~100 (±2% tolerance for rounding).
        - kcal values must be consistent with grams (P/C: 4 kcal/g, F: 9 kcal/g).
        - All values must be non-negative.
        Violations raise ValueError so upstream bugs surface immediately
        instead of producing silent inconsistencies downstream.
        """
        for name, val in [
            ("protein_g", self.protein_g), ("fat_g", self.fat_g), ("carb_g", self.carb_g),
            ("protein_pct", self.protein_pct), ("fat_pct", self.fat_pct), ("carb_pct", self.carb_pct),
            ("protein_kcal", self.protein_kcal), ("fat_kcal", self.fat_kcal), ("carb_kcal", self.carb_kcal),
        ]:
            if val < 0:
                raise ValueError(f"MacroSplit.{name} must be non-negative, got {val}")
        pct_sum = self.protein_pct + self.fat_pct + self.carb_pct
        if abs(pct_sum - 100.0) > 2.0:
            raise ValueError(
                f"MacroSplit percentages must sum to ~100, got {pct_sum:.1f} "
                f"(P={self.protein_pct}, F={self.fat_pct}, C={self.carb_pct})"
            )

    @property
    def total_kcal(self) -> float:
        return self.protein_kcal + self.fat_kcal + self.carb_kcal

    @property
    def carbs_clamped(self) -> bool:
        """Phase-6: True if carbs were clamped to 0 because protein+fat exceeded target.

        Phase-6 fix: previously scanned notes for "carb" AND "clamp", but the
        actual note emitted by `compute_carbs` said "Carbs set to 0" (not
        "clamped"). Now we also accept "set to 0" as the clamped signal, AND
        we defensively check `carb_g == 0` (the most reliable signal).
        """
        if self.carb_g == 0:
            return True
        return any(
            "carb" in note.lower() and ("clamp" in note.lower() or "set to 0" in note.lower())
            for note in self.notes
        )


@dataclass
class HydrationTarget:
    water_liters_per_day: float
    components: dict = field(default_factory=dict)
    notes: list[str] = field(default_factory=list)


@dataclass
class MicronutrientTargets:
    fiber_g: float
    fruit_cups: int
    veg_cups: int
    notes: list[str] = field(default_factory=list)


@dataclass
class NutritionPlan:
    rmr: RMRResult
    tdee: TDEEResult
    calories: CalorieTargets
    macros: MacroSplit
    hydration: HydrationTarget
    micronutrients: MicronutrientTargets
    timeline_weeks: int                            # estimated time to goal
    notes: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return convert_for_json(self)


__all__ = [
    "CalorieStrategy",
    "RMRFormula",
    "RMRResult",
    "TDEEResult",
    "CalorieTargets",
    "MacroSplit",
    "HydrationTarget",
    "MicronutrientTargets",
    "NutritionPlan",
]

File: download/fitn/fitn_engine/lib/src/models/nutrition.dart
/// Nutrition plan output models. See spec §9.1 (output tree).
library;

import 'enums.dart';

class RMRResult {
  RMRResult({
    required this.formula,
    required this.baseRmrKcal,
    required this.metabolicAdaptationFactor,
    required this.weightReducedFactor,
    required this.adjustedRmrKcal,
    required this.notes,
  });

  final RMRFormula formula;
  final double baseRmrKcal;
  final double metabolicAdaptationFactor;
  final double weightReducedFactor;
  final double adjustedRmrKcal;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'formula': formula.toJson(),
        'base_rmr_kcal': baseRmrKcal,
        'metabolic_adaptation_factor': metabolicAdaptationFactor,
        'weight_reduced_factor': weightReducedFactor,
        'adjusted_rmr_kcal': adjustedRmrKcal,
        'notes': notes,
      };

  factory RMRResult.fromJson(Map<String, dynamic> json) {
    return RMRResult(
      formula: RMRFormulaJson.fromJson(json['formula'] as String),
      baseRmrKcal: (json['base_rmr_kcal'] as num).toDouble(),
      metabolicAdaptationFactor:
          (json['metabolic_adaptation_factor'] as num).toDouble(),
      weightReducedFactor:
          (json['weight_reduced_factor'] as num).toDouble(),
      adjustedRmrKcal: (json['adjusted_rmr_kcal'] as num).toDouble(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class TDEEResult {
  TDEEResult({
    required this.rmrKcal,
    required this.activityFactor,
    required this.tdeeKcal,
    this.adaptiveTdeeKcal,
    required this.finalTdeeKcal,
    required this.notes,
  });

  final double rmrKcal;
  final double activityFactor;
  final double tdeeKcal;
  final double? adaptiveTdeeKcal;
  final double finalTdeeKcal;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'rmr_kcal': rmrKcal,
        'activity_factor': activityFactor,
        'tdee_kcal': tdeeKcal,
        'adaptive_tdee_kcal': adaptiveTdeeKcal,
        'final_tdee_kcal': finalTdeeKcal,
        'notes': notes,
      };

  factory TDEEResult.fromJson(Map<String, dynamic> json) {
    return TDEEResult(
      rmrKcal: (json['rmr_kcal'] as num).toDouble(),
      activityFactor: (json['activity_factor'] as num).toDouble(),
      tdeeKcal: (json['tdee_kcal'] as num).toDouble(),
      adaptiveTdeeKcal: (json['adaptive_tdee_kcal'] as num?)?.toDouble(),
      finalTdeeKcal: (json['final_tdee_kcal'] as num).toDouble(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class CalorieTargets {
  CalorieTargets({
    required this.strategy,
    required this.baseTdeeKcal,
    required this.ratePct,
    required this.rateLabel,
    required this.calorieDeltaKcal,
    required this.targetCaloriesKcal,
    required this.calorieFloorApplied,
    this.floorKcal,
    required this.notes,
  });

  final CalorieStrategy strategy;
  final double baseTdeeKcal;
  final double ratePct;
  final String rateLabel;
  final double calorieDeltaKcal;
  final double targetCaloriesKcal;
  final bool calorieFloorApplied;
  final double? floorKcal;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'strategy': strategy.toJson(),
        'base_tdee_kcal': baseTdeeKcal,
        'rate_pct': ratePct,
        'rate_label': rateLabel,
        'calorie_delta_kcal': calorieDeltaKcal,
        'target_calories_kcal': targetCaloriesKcal,
        'calorie_floor_applied': calorieFloorApplied,
        'floor_kcal': floorKcal,
        'notes': notes,
      };

  factory CalorieTargets.fromJson(Map<String, dynamic> json) {
    return CalorieTargets(
      strategy: CalorieStrategyJson.fromJson(json['strategy'] as String),
      baseTdeeKcal: (json['base_tdee_kcal'] as num).toDouble(),
      ratePct: (json['rate_pct'] as num).toDouble(),
      rateLabel: json['rate_label'] as String,
      calorieDeltaKcal: (json['calorie_delta_kcal'] as num).toDouble(),
      targetCaloriesKcal: (json['target_calories_kcal'] as num).toDouble(),
      calorieFloorApplied: json['calorie_floor_applied'] as bool,
      floorKcal: (json['floor_kcal'] as num?)?.toDouble(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class MacroSplit {
  MacroSplit({
    required this.proteinG,
    required this.fatG,
    required this.carbG,
    required this.proteinPct,
    required this.fatPct,
    required this.carbPct,
    required this.proteinKcal,
    required this.fatKcal,
    required this.carbKcal,
    required this.notes,
  });

  final double proteinG;
  final double fatG;
  final double carbG;
  final double proteinPct;
  final double fatPct;
  final double carbPct;
  final double proteinKcal;
  final double fatKcal;
  final double carbKcal;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'protein_g': proteinG,
        'fat_g': fatG,
        'carb_g': carbG,
        'protein_pct': proteinPct,
        'fat_pct': fatPct,
        'carb_pct': carbPct,
        'protein_kcal': proteinKcal,
        'fat_kcal': fatKcal,
        'carb_kcal': carbKcal,
        'notes': notes,
      };

  factory MacroSplit.fromJson(Map<String, dynamic> json) {
    return MacroSplit(
      proteinG: (json['protein_g'] as num).toDouble(),
      fatG: (json['fat_g'] as num).toDouble(),
      carbG: (json['carb_g'] as num).toDouble(),
      proteinPct: (json['protein_pct'] as num).toDouble(),
      fatPct: (json['fat_pct'] as num).toDouble(),
      carbPct: (json['carb_pct'] as num).toDouble(),
      proteinKcal: (json['protein_kcal'] as num).toDouble(),
      fatKcal: (json['fat_kcal'] as num).toDouble(),
      carbKcal: (json['carb_kcal'] as num).toDouble(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class HydrationTarget {
  HydrationTarget({
    required this.waterLitersPerDay,
    required this.components,
    required this.notes,
  });

  final double waterLitersPerDay;
  final Map<String, double> components;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'water_liters_per_day': waterLitersPerDay,
        'components': components,
        'notes': notes,
      };

  factory HydrationTarget.fromJson(Map<String, dynamic> json) {
    return HydrationTarget(
      waterLitersPerDay: (json['water_liters_per_day'] as num).toDouble(),
      components: (json['components'] as Map).map(
          (k, v) => MapEntry(k as String, (v as num).toDouble())),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class MicronutrientTargets {
  MicronutrientTargets({
    required this.fiberG,
    required this.fruitCups,
    required this.vegCups,
    required this.notes,
  });

  final double fiberG;
  final int fruitCups;
  final int vegCups;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'fiber_g': fiberG,
        'fruit_cups': fruitCups,
        'veg_cups': vegCups,
        'notes': notes,
      };

  factory MicronutrientTargets.fromJson(Map<String, dynamic> json) {
    return MicronutrientTargets(
      fiberG: (json['fiber_g'] as num).toDouble(),
      fruitCups: (json['fruit_cups'] as num).toInt(),
      vegCups: (json['veg_cups'] as num).toInt(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class NutritionPlan {
  NutritionPlan({
    required this.rmr,
    required this.tdee,
    required this.calories,
    required this.macros,
    required this.hydration,
    required this.micronutrients,
    required this.timelineWeeks,
    required this.notes,
  });

  final RMRResult rmr;
  final TDEEResult tdee;
  final CalorieTargets calories;
  final MacroSplit macros;
  final HydrationTarget hydration;
  final MicronutrientTargets micronutrients;
  final int timelineWeeks;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'rmr': rmr.toJson(),
        'tdee': tdee.toJson(),
        'calories': calories.toJson(),
        'macros': macros.toJson(),
        'hydration': hydration.toJson(),
        'micronutrients': micronutrients.toJson(),
        'timeline_weeks': timelineWeeks,
        'notes': notes,
      };

  factory NutritionPlan.fromJson(Map<String, dynamic> json) {
    return NutritionPlan(
      rmr: RMRResult.fromJson(json['rmr'] as Map<String, dynamic>),
      tdee: TDEEResult.fromJson(json['tdee'] as Map<String, dynamic>),
      calories:
          CalorieTargets.fromJson(json['calories'] as Map<String, dynamic>),
      macros: MacroSplit.fromJson(json['macros'] as Map<String, dynamic>),
      hydration:
          HydrationTarget.fromJson(json['hydration'] as Map<String, dynamic>),
      micronutrients: MicronutrientTargets.fromJson(
          json['micronutrients'] as Map<String, dynamic>),
      timelineWeeks: (json['timeline_weeks'] as num).toInt(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}
